accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,221 @@
+
+Another way is to invest in [a proposed runtime feature](https://github.com/dotnet/runtime/issues/10480) which would make the GC more aggressive about zeroing memory on compaction. This has the benefit that _many_ different types – including standard strings – would become candidates for having their data zeroed out during a compaction phase. High-sensitivity applications may wish to opt in to this GC mode anyway depending on their risk analysis.
+
+### Minimizing misuse of sensitive data
+
+Here, the goal isn't to hide secrets from memory, but to signal to developers "this data is sensitive and should be handled cautiously." In other words, the goal is to minimize the risk of a developer _accidentally_ using the data in an insecure fashion, such as writing it to a log file. SecureString actually does a fine job of this by virtue of the fact that callers must drop down to unsafe code or use the `Marshal` class to interact with it. However, the restrictive SecureString API surface and the fact that we force use of `Marshal` can lead to developers introducing unmanaged memory errors into their applications.
+
+There is opportunity for .NET to provide a better type for this scenario, described in more detail later.
+
+### Using it because it has "secure" in the name
+
+Based both on internal and external conversations, there seems to be a significant audience using the SecureString type because it has "secure" in its name. I have seen many code bases which have contorted themselves into using SecureString in as many places as possible in order to fulfill of a well-intentioned but ill-executed "security compliance" edict. Since this sort of maneuvering generally involves unsafe code, this often ends up _introducing_ security bugs. (One partner was even using unsafe code to zero-out string instances after converting them to SecureString.)
+
+During these conversations it's apparent that they're not attempting to mitigate any specific threats. They're instead using SecureString because it's listed as a requirement in their engineering processes. These users would be best served by using the existing string class or by using the `NetworkCredential` wrapper as suggested earlier.
+
+## A proposal for marking data as sensitive
+
+Of all the scenarios mentioned above, the one that strikes me as most interesting is the use of SecureString as a marker class to mean "you should be careful when using this data." It's possible to extend this into a more generalized proposal that can serve as a wrapper for arbitary types _T_, so it can be used to hold both `byte` and `char`-based buffers.
+
+```cs
+namespace System.Security
+{
+    public sealed class SensitiveData<T> : IDisposable where T : unmanaged
+    {
+        public SensitiveData(ReadOnlySpan<T> contents);
+        public int Length { get; }
+        public void CopyTo(Span<T> destination);
+        public void Dispose();
+    }
+}
```

Comment:
Does this need to be generic? I'd think practically `T` would only ever be `char` or `byte`, no?

Reply:
Doesn't seem like it would hurt things if it's constrained to be unmanaged like it is now, does it? E.g. someone could potentially use a user-defined struct that satisfies the constraint. I'm just wondering what it would improve if it was constrained to a few pre-chosen types.

Reply:
In theory it could mark any arbitrary buffer as "sensitive". Bytes and chars would be the most common, but perhaps you have a collection of (some object model) where the objects themselves are the sensitive data. We could remove the unmanaged constraint if it really came down to it. It's just that we can't provide some extra defense-in-depth protections (against heap corruption, say) unless the T is unmanaged. But TBH that's just an implementation detail.

Reply:
I'll shortly update the doc with a blurb about the _T_.